src/H5Eint.c

```diff
@@ -98,6 +120,738 @@ char H5E_mpi_error_str[MPI_MAX_ERROR_STRING];
+ * Function:    H5E__get_current_stack
+ *
+ * Purpose:     Private function to register an error stack.
+ *
+ * Return:      Success:    Pointer to an error class struct
+ *              Failure:    NULL
+ *
+ *-------------------------------------------------------------------------
+ */
+H5E_stack_t *
+H5E__get_current_stack(void)
+{
+    H5E_stack_t *current_stack;      /* Pointer to the current error stack */
+    H5E_stack_t *estack_copy = NULL; /* Pointer to new error stack to return */
+    unsigned     u;                  /* Local index variable */
+    H5E_stack_t *ret_value = NULL;   /* Return value */
+
+    FUNC_ENTER_PACKAGE
+
+    /* Get a pointer to the current error stack */
+    if (NULL == (current_stack = H5E__get_my_stack()))
+        HGOTO_ERROR(H5E_ERROR, H5E_CANTGET, NULL, "can't get current error stack");
+
+    /* Allocate a new error stack */
+    if (NULL == (estack_copy = H5FL_CALLOC(H5E_stack_t)))
+        HGOTO_ERROR(H5E_RESOURCE, H5E_NOSPACE, NULL, "memory allocation failed");
+
+    /* Make a copy of current error stack */
+    estack_copy->nused = current_stack->nused;
+    for (u = 0; u < current_stack->nused; u++)
+        if (H5E__copy_stack_entry(&estack_copy->entries[u], &current_stack->entries[u]) < 0)
+            HGOTO_ERROR(H5E_ERROR, H5E_CANTSET, NULL, "can't set error entry");
+
+    /* Copy the "automatic" error reporting information */
+    estack_copy->auto_op   = current_stack->auto_op;
+    estack_copy->auto_data = current_stack->auto_data;
+
+    /* Empty current error stack */
+    H5E__clear_stack(current_stack);
+
+    /* Set the return value */
+    ret_value = estack_copy;
+
+done:
+    if (ret_value == NULL)
+        if (estack_copy)
+            estack_copy = H5FL_FREE(H5E_stack_t, estack_copy);
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5E__get_current_stack() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5E__set_current_stack
+ *
+ * Purpose:     Private function to replace an error stack.
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5E__set_current_stack(H5E_stack_t *estack)
+{
+    H5E_stack_t *current_stack;       /* Default error stack */
+    unsigned     u;                   /* Local index variable */
+    herr_t       ret_value = SUCCEED; /* Return value */
+
+    FUNC_ENTER_PACKAGE
+
+    /* Sanity check */
+    assert(estack);
+
+    /* Get a pointer to the current error stack */
+    if (NULL == (current_stack = H5E__get_my_stack()))
+        HGOTO_ERROR(H5E_ERROR, H5E_CANTGET, FAIL, "can't get current error stack");
+
+    /* Empty current error stack */
+    H5E__clear_stack(current_stack);
+
+    /* Copy new stack to current error stack */
+    current_stack->nused = estack->nused;
+    for (u = 0; u < current_stack->nused; u++)
+        if (H5E__copy_stack_entry(&current_stack->entries[u], &estack->entries[u]) < 0)
```

Comment:
Refactored all the routines that manipulate stack entries to use a new, common stack entry copy routine.